src/NuGet.Core/NuGet.Packaging/Signing/Signatures/AlgorithmIdentifier.cs

```diff
@@ -0,0 +1,49 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Security.Cryptography;
+using NuGet.Packaging.Signing.DerEncoding;
+
+namespace NuGet.Packaging.Signing
+{
+    /*
+        From RFC 3280 (https://tools.ietf.org/html/rfc3280#section-4.1.1.2):
+
+            AlgorithmIdentifier ::= SEQUENCE {
+                algorithm               OBJECT IDENTIFIER,
+                parameters              ANY DEFINED BY algorithm OPTIONAL
+            }
+    */
+    /// <remarks>This is public only to facilitate testing.</remarks>
+    public sealed class AlgorithmIdentifier
+    {
+        public Oid Algorithm { get; }
+
+        internal AlgorithmIdentifier(Oid algorithm)
+        {
+            Algorithm = algorithm;
+        }
+
+        public static AlgorithmIdentifier Read(byte[] bytes)
+        {
+            var reader = DerSequenceReader.CreateForPayload(bytes);
+
+            return Read(reader);
+        }
+
+        internal static AlgorithmIdentifier Read(DerSequenceReader reader)
+        {
+            var algIdReader = reader.ReadSequence();
+            var algorithm = algIdReader.ReadOid();
+
+            // Skip the "parameters" field.  We do not use it.
```

Comment:
Won't this leave the reader in a bad state if we don't consume the `parameters`? If this method was private it wouldn't be an issue since the only entry point provides a byte array.

Reply:
The short answer is no.

`reader.ReadSequence()` consumes the bytes for the sequence and passes them to a new `DerSequenceReader`.  The original reader will be in a good state with its position after the sequence.  The new reader is discarded after reading the algorithm OID.